include/mbedtls/pk_info.h

```diff
@@ -0,0 +1,274 @@
+                        const unsigned char *hash, size_t hash_len,
+                        const unsigned char *sig, size_t sig_len );
+
+    /** Make signature
+     *
+     * mbedtls_pk_sign() calls this function.
+     *
+     * Assumes that the buffer \c sig has room for
+     * \c signature_size_func(ctx) bytes.
+     *
+     * The arguments \c f_rng and \c p_rng are provided in case the
+     * algorithm requires randomization. Implementations are not
+     * required to use it if they have their own random source. If \c
+     * f_rng is null, the implementation should operate if it can, and
+     * return #MBEDTLS_ERR_PK_BAD_INPUT_DATA otherwise.
+     *
+     * Opaque implementations may omit this method if they do not support
+     * signing. */
+    int (*sign_func)( void *ctx, mbedtls_md_type_t md_alg,
+                      const unsigned char *hash, size_t hash_len,
+                      unsigned char *sig, size_t *sig_len,
+                      int (*f_rng)(void *, unsigned char *, size_t),
+                      void *p_rng );
+
+    /** Decrypt message
+     *
+     * mbedtls_pk_decrypt() calls this function.
+     *
+     * The arguments \c f_rng and \c p_rng are provided in case the
+     * algorithm requires randomization. Implementations are not
+     * required to use it if they have their own random source. If \c
+     * f_rng is null, the implementation should operate if it can, and
+     * return #MBEDTLS_ERR_PK_BAD_INPUT_DATA otherwise.
+     *
+     * Opaque implementations may omit this method if they do not support
+     * decryption. */
+    int (*decrypt_func)( void *ctx, const unsigned char *input, size_t ilen,
+                         unsigned char *output, size_t *olen, size_t osize,
+                         int (*f_rng)(void *, unsigned char *, size_t),
+                         void *p_rng );
+
+    /** Encrypt message
+     *
+     * mbedtls_pk_decrypt() calls this function.
+     *
+     * The arguments \c f_rng and \c p_rng are provided in case the
+     * algorithm requires randomization. Implementations are not
+     * required to use it if they have their own random source. If \c
+     * f_rng is null, the implementation should operate if it can, and
+     * return #MBEDTLS_ERR_PK_BAD_INPUT_DATA otherwise.
+     *
+     * Opaque implementations may omit this method if they do not support
+     * encryption. */
+    int (*encrypt_func)( void *ctx, const unsigned char *input, size_t ilen,
+                         unsigned char *output, size_t *olen, size_t osize,
+                         int (*f_rng)(void *, unsigned char *, size_t),
+                         void *p_rng );
+
+    /** Check public-private key pair
+     *
+     * mbedtls_pk_check_pair() calls this function on the private key pair
+     * object \c prv. The other argument \c pub may be of any type, but it
+     * is guaranteed to be initialized.
+     *
+     * Opaque implementations may omit this method. */
```

Comment:
@mpg @hanno-arm @yanesca Final thoughts on whether to make this method mandatory?

Also we should explain what the minimum support is. Proposal, if we decide that this method is mandatory:

> * If `\c prv` is an RSA key and `\c pub` is a transparent RSA key (i.e. `\c pub` has the type `#MBEDTLS_PK_RSA` or `#MBEDTLS_PK_RSASSA`), then `\c check_pair_func` must return 0 if the public key is mathematically equivalent to the public part of `\c prv`, and `#MBEDTLS_ERR_RSA_KEY_CHECK_FAILED` otherwise.
> * If `\c pub` is an ECC key and `\c pub` is a transparent ECC key that can be used for ECDSA (i.e. `\c pub` has the type `#MBEDTLS_PK_ECKEY` or `#MBEDTLS_PK_ECDSA`), then `check_pair_func` must return 0 if the public key is mathematically equivalent to the public part of `prv`, and `MBEDTLS_ERR_ECP_BAD_INPUT_DATA` otherwise.
> * If `\c pub` is a transparent key (key of type `#MBEDTLS_PK_RSA`, `#MBEDTLS_PK_RSASSA`, `#MBEDTLS_PK_ECKEY` or `#MBEDTLS_PK_ECDSA`) whose type does not match the semantic type of `\c prv` (RSA, ECC or other), then `check_pair_func` must return `#MBEDTLS_ERR_PK_TYPE_MISMATCH`.
> * If `\c pub` and `\c prv` are opaque keys from the same engines (i.e. ``), then `check_pair_func` must return 0, `#MBEDTLS_ERR_PK_TYPE_MISMATCH`, or `#MBEDTLS_ERR_RSA_KEY_CHECK_FAILED` or `#MBEDTLS_ERR_ECP_BAD_INPUT_DATA`  as in the case of transparent keys.
> * If `\c pub` is an opaque key which is not from the same engine as `\c prv`, then `check_pair_func` may either return a semantically correct status as in the case of transparent keys, or it may return `#MBEDTLS_ERR_PK_FEATURE_UNAVAILABLE`.
> * Alternatively, `check_pair_func` may return another PK, RSA or ECP error code if applicable.

Reply:
I don't have a strong opinion here, but in principle I'd like to make it mandatory just to make opaque closer to non-opaque from a user's perspective. Also, since the goal is that the user can gain some confidence that the provisioned private key and certificate match, I think this is relevant for opaque keys.

I agree with the proposed documentation, but I'd like to add to it that the implementation is allowed to do a probabilistic and computationally expensive check (for example do a private operation on some data, then a public operation and check that "it works"). Users should be aware that this function may be expensive and avoid calling it more often than necessary (it's already the case for RSA_ALT, unfortunately it wasn't documented so far, so this should be added to the public documentation in `pk.h`).

Reply:
I would prefer to make it mandatory too and the proposed documentation looks good to me.

Reply:
Done